File: shared/scripts/js/nvim-fzf-preview.py
#!/usr/bin/env python3
import subprocess
import sys
import os
from typing import List, Dict
import shutil


def main():
    """Main function that orchestrates the file finding and editing process."""
    try:
        config = get_script_config()
        deps = check_dependencies()

        try:
            if deps["has_fd"]:
                result = find_files_with_fd(config, deps)
            else:
                result = find_files_with_find(config, deps)

            selected_file = result.strip()
            if selected_file:
                subprocess.run(["nano", selected_file], check=True)
            else:
                print("No file selected.")
                sys.exit(1)

        except (subprocess.CalledProcessError, KeyboardInterrupt):
            print("No file selected.")
            sys.exit(1)

    except Exception as error:
        print(f"An unexpected error occurred: {error}")
        sys.exit(1)


def get_script_config() -> Dict[str, any]:
    """Get configuration from command line arguments and environment variables."""
    search_path = sys.argv[1] if len(sys.argv) > 1 else None

    project_dir = os.environ.get("PROJECT_DIR", ".")
    work_dir = os.environ.get("WORK_DIR", ".")
    asset_dir = os.environ.get("ASSET_DIR", ".")

    return {
        "search_path": search_path,
        "default_search_dirs": [project_dir, work_dir, asset_dir],
    }


def command_exists(command: str) -> bool:
    """Check if a command exists in the system PATH."""
    return shutil.which(command) is not None


def check_dependencies() -> Dict[str, bool]:
    """Check for the availability of optional dependencies."""
    return {
        "has_fd": command_exists("fd"),
        "has_bat": command_exists("bat"),
        "has_fzf": command_exists("fzf"),
    }


def build_fzf_args(has_bat: bool) -> List[str]:
    """Build fzf arguments based on available dependencies."""
    if has_bat:
        return [
            "--ansi",
            "--preview-window",
            "right:45%",
            "--preview",
            "bat --color=always --style=header,grid --line-range :300 {}",
        ]
    return []


def find_files_with_fd(config: Dict[str, any], deps: Dict[str, bool]) -> str:
    """Find files using fd command."""
    paths_to_search = (
        [config["search_path"]]
        if config["search_path"]
        else config["default_search_dirs"]
    )

    # Build fd command
    fd_args = ["fd", "--type", "f"]
    for path in paths_to_search:
        fd_args.extend(["--search-path", path])
    fd_args.extend(["--follow", "--hidden", "--exclude", ".git"])

    # Build fzf command
    fzf_args = ["fzf"] + build_fzf_args(deps["has_bat"])

    # Run fd piped to fzf
    fd_process = subprocess.Popen(fd_args, stdout=subprocess.PIPE, text=True)
    fzf_process = subprocess.Popen(
        fzf_args, stdin=fd_process.stdout, stdout=subprocess.PIPE, text=True
    )

    # Close fd stdout in parent process
    fd_process.stdout.close()

    # Get the result
    result, _ = fzf_process.communicate()

    # Wait for fd to complete and check return codes
    fd_process.wait()
    if fzf_process.returncode != 0:
        raise subprocess.CalledProcessError(fzf_process.returncode, "fzf")

    return result


def find_files_with_find(config: Dict[str, any], deps: Dict[str, bool]) -> str:
    """Find files using find command."""
    paths_to_search = (
        [config["search_path"]]
        if config["search_path"]
        else config["default_search_dirs"]
    )

    # Build find command
    find_args = ["find"] + paths_to_search + ["-type", "f"]

    # Build fzf command
    fzf_args = ["fzf"] + build_fzf_args(deps["has_bat"])

    # Run find piped to fzf
    find_process = subprocess.Popen(find_args, stdout=subprocess.PIPE, text=True)
    fzf_process = subprocess.Popen(
        fzf_args, stdin=find_process.stdout, stdout=subprocess.PIPE, text=True
    )

    # Close find stdout in parent process
    find_process.stdout.close()

    # Get the result
    result, _ = fzf_process.communicate()

    # Wait for find to complete and check return codes
    find_process.wait()
    if fzf_process.returncode != 0:
        raise subprocess.CalledProcessError(fzf_process.returncode, "fzf")

    return result


if __name__ == "__main__":
    main()

File: shared/scripts/deprecated/js/nvim-fzf-preview.ts
#!/usr/bin/env bun
import { $ } from "bun";

main().catch((error) => {
  console.error('An unexpected error occurred:', error.message);
  process.exit(1);
});

async function main() {
  const config = getScriptConfig();
  const deps = await checkDependencies();

  try {
    let result: string;

    if (deps.hasFd) {
      const pathsToSearch = config.searchPath ? [config.searchPath] : config.defaultSearchDirs;
      const fdArgs = pathsToSearch.flatMap(p => ['--search-path', p]);
      const fileListCmd = ['fd', '--type', 'f', ...fdArgs, '--follow', '--hidden', '--exclude', '.git'];
      const fzfArgs = buildFzfArgs(deps.hasBat);

      result = await $`${fileListCmd} | fzf ${fzfArgs}`.text();
    } else {
      const pathsToSearch = config.searchPath ? [config.searchPath] : config.defaultSearchDirs;
      const fzfArgs = buildFzfArgs(deps.hasBat);

      result = await $`find ${pathsToSearch} -type f | fzf ${fzfArgs}`.text();
    }

    const selectedFile = result.trim();
    if (selectedFile) {
      await $`nano ${selectedFile}`;
    } else {
      console.log('No file selected.');
      process.exit(1);
    }
  } catch (error) {
    console.log('No file selected.');
    process.exit(1);
  }
}

function getScriptConfig(): { searchPath: string; defaultSearchDirs: string[]; } {
  const searchPath = process.argv[2];
  const PROJECT_DIR = process.env.PROJECT_DIR || '.';
  const WORK_DIR = process.env.WORK_DIR || '.';
  const ASSET_DIR = process.env.ASSET_DIR || '.';

  return {
    searchPath,
    defaultSearchDirs: [PROJECT_DIR, WORK_DIR, ASSET_DIR],
  };
}

async function commandExists(command: string): Promise<boolean> {
  try {
    await $`command -v ${command}`.quiet();
    return true;
  } catch {
    return false;
  }
}

async function checkDependencies(): Promise<{ hasFd: boolean; hasBat: boolean; }> {
  const [hasFd, hasBat] = await Promise.all([
    commandExists('fd'),
    commandExists('bat'),
  ]);

  return { hasFd, hasBat };
}

function buildFzfArgs(hasBat: boolean): string[] {
  if (hasBat) {
    return [
      '--ansi',
      '--preview-window', 'right:45%',
      '--preview', 'bat --color=always --style=header,grid --line-range :300 {}'
    ];
  }
  return [];
}
